src/v/raft/seq_stm.cc

```diff
@@ -0,0 +1,337 @@
+                raftlog.trace,
+                "seq_stm snapshot with offset:{} is loaded",
+                data.offset);
+          });
+    });
+}
+
+ss::future<> seq_stm::persist_snapshot(iobuf&& data) {
+    seq_snapshot_header header;
+    header.version = seq_snapshot_header::supported_version;
+    header.snapshot_size = data.size_bytes();
+
+    iobuf data_size_buf;
+    reflection::serialize(data_size_buf, header.version, header.snapshot_size);
+
+    return _snapshot_mgr.start_snapshot().then(
+      [this, data = std::move(data), data_size_buf = std::move(data_size_buf)](
+        storage::snapshot_writer writer) mutable {
+          return ss::do_with(
+            std::move(writer),
+            [this,
+             data = std::move(data),
+             data_size_buf = std::move(data_size_buf)](
+              storage::snapshot_writer& writer) mutable {
+                return writer.write_metadata(std::move(data_size_buf))
+                  .then([&writer, data = std::move(data)]() mutable {
+                      return write_iobuf_to_output_stream(
+                        std::move(data), writer.output());
+                  })
+                  .finally([&writer] { return writer.close(); })
+                  .then([this, &writer] {
+                      return _snapshot_mgr.finish_snapshot(writer);
+                  });
+            });
+      });
+}
+
+ss::future<> seq_stm::make_snapshot() {
+    return _op_lock.with([this]() {
+        auto f = ss::now();
+        if (!_resolved_when_snapshot_hydrated.available()) {
+            f = _resolved_when_snapshot_hydrated.get_shared_future();
+        }
+
+        return f.then([this] { return do_make_snapshot(); });
+    });
+}
+
+ss::future<> seq_stm::ensure_snapshot_exists(model::offset target_offset) {
+    return _op_lock.with([this, target_offset]() {
+        auto f = ss::now();
+        if (!_resolved_when_snapshot_hydrated.available()) {
+            f = _resolved_when_snapshot_hydrated.get_shared_future();
+        }
+
+        return f.then([this, target_offset] {
+            if (target_offset <= _last_snapshot_offset) {
+                return ss::now();
+            } else {
+                return wait(target_offset, model::no_timeout)
+                  .then([this, target_offset]() {
+                      vassert(
+                        target_offset <= _insync_offset,
+                        "after we waited for target_offset ({}) _insync_offset "
+                        "({}) should have matched it or bypassed",
+                        target_offset,
+                        _insync_offset);
+                      return do_make_snapshot();
+                  });
+            }
+        });
+    });
+}
+
+ss::future<> seq_stm::do_make_snapshot() {
+    vlog(raftlog.trace, "saving seq_stm snapshot");
+    snapshot data;
+    data.offset = _insync_offset;
+    for (auto& entry : _seq_table) {
+        data.entries.push_back(entry.second);
```

Comment:
do you intent to make this copy?

Reply:
Yes, I'm making a copy of the current state to save it as a snapshot to disk. Do you think there is a better way?

Reply:
this makes sense to me, just wanted to make sure this was intended.